server/src/main/java/org/elasticsearch/cluster/coordination/MasterHistory.java

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.coordination;
+
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.core.Nullable;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+/**
+ * This class represents a node's view of the history of which nodes have been elected master over the last 30 minutes. It is kept in
+ * memory, so when a node comes up it does not have any knowledge of previous master history before that point.
+ */
+public class MasterHistory implements ClusterStateListener, Writeable, Writeable.Reader<MasterHistory> {
+    private List<TimeAndMaster> masterHistory;
+    Supplier<Long> nowSupplier = System::currentTimeMillis; // Can be changed for testing
```

Comment:
Since we're working with "things that happened since / in the last X minutes/seconds" we should use `nanoTime` here as it's monotonically increasing

Reply:
I recommend `ThreadPool#relativeTimeInMillis` which is monotonic but easily faked in tests.

Reply:
I actually think what I did isn't going to work at all the way I've done it if we don't have somewhat-in-sync times across boxes. The reason is the `hasSeenMasterInLastNSeconds` method. If the times are all only meaningful on the current machine then there's no way to know if we've seen a master in the last 30 seconds on a remote machine, unless we calculate that before we send it over the wire. Might need to rethink this one.

Reply:
Would it be better to have each of these methods be its own remote call rather than returning a MasterHistory object? I'm stuck on how to answer `hasSeenMasterInLastNSeconds` without it being its own call. And even then, it could easily take more than 30 seconds to respond, so the answer would be void before we even got it.

Reply:
>if we don't have somewhat-in-sync times across boxes

This is an invariant we have to work with (without specialised hardware we can't do much about it)

I don't completely understand why you'd need to measure and compare time across machines though? Can you maybe explain that?

Machine 1 - computes its local history (ie. masters seen in last X seconds in relation to an origin chosen on machine 1)
Machine 2 - computes its own local history (ie. masters seen in the last X seconds in relation to an origin chosen on machine 2)

Machine 1 might ask Machine 2 about its history, in which case Machine 2 will reply with the history it has in its local "cache" for the last X seconds, according to the origin chosen on Machine 2 (no comparison to Machine 1 time/origin). Machine 1 will receive this history and if it contains repeated changes or master going null/not-null Machine 1 will be able to conclude that Machine 2 is an unstable master (again, Master 1 will work with the entire history it received from Machine 2, without comparing the times to its local seen masters, nor truncating/purging it in any way).

Reply:
The reason is that in machine 2's history there is no notion of "now". There are just a few events with timestamps that are meaningless to machine 1. Machine 1 can't tell if the newest even was 2 seconds or 2 years ago. I'll change MasterHistory to ship some notion of "now" so that i can answer questions like "has there been a master in the last 30 seconds". I won't need to expose what "now" is -- it'll just be some anchor to answer relative questions like that.

Reply:
>The reason is that in machine 2's history there is no notion of "now".

There is, but it's relative (and local) to Machine 2.

>Machine 1 can't tell if the newest even was 2 seconds or 2 years ago. 

Maybe I'm missing something but each machine will only hold 30 minutes' worth of history AFAIK

Reply:
OK I broke it into a MutableMasterHistory that is updated live on the local node, and an ImmutableMasterHistory that is never updated but that can be passed from one node to another.